docs/fundamentals/code-analysis/quality-rules/ca1869.md

```diff
@@ -0,0 +1,122 @@
+---
+title: "CA1869: Cache and reuse 'JsonSerializerOptions' instances"
+description: "Learn about code analyzer rule CA1869 - Cache and reuse 'JsonSerializerOptions' instances"
+ms.date: 08/21/2023
+ms.topic: reference
+f1_keywords:
+ - CA1869
+ - AvoidSingleUseOfLocalJsonSerializerOptions
+helpviewer_keywords:
+ - CA1869
+author: Jozkee
+ms.author: dacantu
+---
+
+# CA1869: Cache and reuse 'JsonSerializerOptions' instances
+
+|                                     | Value                                  |
+| ----------------------------------- |----------------------------------------|
+| **Rule ID**                         | CA1869                                 |
+| **Category**                        | [Performance](performance-warnings.md) |
+| **Fix is breaking or non-breaking** | Non-breaking                           |
+| **Enabled by default in .NET 7**    | No                                     |
+
+## Cause
+
+A local instance of <xref:System.Text.Json.JsonSerializerOptions> is used once as the `options` argument of a <xref:System.Text.Json.JsonSerializer.Serialize> or <xref:System.Text.Json.JsonSerializer.Deserialize> call.
+
+## Rule description
+
+Using a local instance of <xref:System.Text.Json.JsonSerializerOptions> for serialization or deserialization can substantially degrade the performance of your application if your code executes multiple times since System.Text.Json internally caches serialization releated metadata into the provided instance.
+
+## How to fix violations
+
+You can use the singleton pattern to avoid creating a new <xref:System.Text.Json.JsonSerializerOptions> instance every time your code is executed.
+
+## Example
+
+The following code snippet shows two violations of CA1869:
+
+```csharp
+static string Serialize<T>(T value)
+{
+    JsonSerializerOptions jsonOptions = new() 
+    { 
+        WriteIndented = true
+    };
+
+    return JsonSerializer.Serialize(value, jsonOptions);
+}
+
+static T Deserialize<T>(string json)
+{
+    return JsonSerializer.Deserialize(value, new JsonSerializeOptions { AllowTrailingCommas = true });
+}
+```
+
+The following code snippet fixes the violation:
+
+```csharp
+private static readonly JsonSerializerOptions s_jsonOptions = new()
+{
+    WriteIndented = true,
+    AllowTrailingCommas = true
+};
+
+static string Serialize<T>(T value)
+{
+    return JsonSerializer.Serialize(value, s_jsonOptions);
+}
+
+static T Deserialize<T>(string json)
+{
+    return JsonSerializer.Deserialize(value, s_jsonOptions);
+}
+```
+
+NOTE: in this case we can safely collapse both `JsonSerializerOptions` into one instance because `WriteIndented` acts on serialization and `AllowTrailingCommas` acts on deserialization and hence not affecting the result of both operations. Do not do this if you need to preserve different behaviors on each operation, instead create separate instances while keeping them global for caching purposes e.g:
```

Comment:
```suggestion
In the previous example, you can safely collapse both `JsonSerializerOptions` instances into one instance. That's because `WriteIndented` acts on serialization and `AllowTrailingCommas` acts on deserialization, hence only the result of one operation is affected. Don't do this if you need to preserve different behaviors on each operation. Instead, create separate instances but still keep them global for caching purposes, for example:
```

Reply:
> hence only the result of one operation is affected.

Not sure if I understand this correctly. What I am trying to express is: in this case you can collapse both instances and the return values of Serialize and Deserialize will not change. 

Is that being expressed on this paragraph?

Reply:
I see what you're trying to say, but I'm wondering if we shouldn't just show two instances to begin with to keep it simpler - one for read and one for write.

Reply:
Yes, we can do that, let me update the example.